### ISSUE TITLE
Force all the reads and writes on "derivative files" (thumbnails and such) to go through the DataAccess framework.

### ISSUE THREAD

**Issue by landreev** [IQSS/dataverse#3764]
In the current implementation we only generate thumbnails for local files. Early on the only non-local files were harvested ones (no physical file - hence no thumbnails). Some direct filesystem read and write calls are being made inside that code. 
Now that we have swift and, potentially, other non-fs storage modes for physical files, these code needs to be updated to go exclusively through the DataAccess subsystem. (which already has functionality to store "derivative files" for the main object - an equivalent of adding files with extra extensions on the filesystem). Once this is done, thumbnails will magically work for swift objects as well. 

The tabular data and export (?) code must be similarly reviewed. (the saving of the generated tabular data file, and preserving the backup of the ingested original is already happening through DataAccess - but needs to be checked and tested w/ swift just in case).

**Comment by djbrooke** [comment]
Due to the pressing nature of the Swift work, this was moved into this sprint.

**Comment by landreev** [comment]
This should go into code review; but let's talk about it during stand up - I feel it needs to be reviewed with more of the team involved.

**Comment by pdurbin** [comment]
@landreev since this is the Swift branch, can you please add to the Installation Guide how to configure `dataverse.files.storage-driver-id` and friends? Here's the writeup I originally posted in https://help.hmdc.harvard.edu/Ticket/Display.html?id=248384 based on a doc written by @ferrys 

For the Swift stuff, here's how my final, working config looks:

```
[root@dvnweb-vm2 ~]# /usr/local/glassfish4/glassfish/bin/asadmin list-jvm-options | grep swift
-Ddataverse.files.storage-driver-id=swift
-Ddataverse.files.swift-endpoint-start=http://rdgw
[root@dvnweb-vm2 ~]#
[root@dvnweb-vm2 ~]# cat /usr/local/glassfish4/glassfish/domains/domain1/config/swift.properties
swift.default.endpoint=endpoint1
swift.auth_url.endpoint1=https://openstack.example.com:35357/v2.0/tokens
swift.tenant.endpoint1=cloud-dataverse
swift.username.endpoint1=REDACTED
swift.password.endpoint1=REDACTED
swift.auth_type.endpoint1=keystone
swift.swift_endpoint.endpoint1=http://rdgw.example.com/swift/v1
[root@dvnweb-vm2 ~]#
```

To get there, I put the file above in place (I assume you'll have your own username and password) and ran the following commands:

`[root@dvnweb-vm2 ~]# /usr/local/glassfish4/glassfish/bin/asadmin create-jvm-options "\-Ddataverse.files.storage-driver-id=swift"`

`[root@dvnweb-vm2 ~]# /usr/local/glassfish4/glassfish/bin/asadmin create-jvm-options "\-Ddataverse.files.swift-endpoint-start=http\://rdgw"`

**Comment by pdurbin** [comment]
At standup this morning I brought up the need to document how to configure Dataverse to use Swift in the Installation Guide. We decided to ask @ferrys to do this in a branch other than `3747-swift-with-derivative-file-support`. @landreev will link her to the right rst file to add a paragraph to. My guess is https://github.com/IQSS/dataverse/blob/develop/doc/sphinx-guides/source/installation/config.rst

**Comment by landreev** [comment]
Sarah is working on the documentation for swift; in the original 3747-swift-and-compute-button branch. I'm going to remove the previous comments, since they were kind of OT in this issue.

**Comment by landreev** [comment]
pull request made: https://github.com/IQSS/dataverse/compare/3747-swift-with-derivative-file-support?expand=1

note: this branch now contains all the code from Sarah; so this branch can/should be used for QA of general swift support. 

Also, note the misleading "3747"  in the branch name! it's the number of the "compute button" issue - which is not ready yet.

**Comment by landreev** [comment]
meant to say "... for QA of general swift support as well"

**Comment by kcondon** [comment]
Couple issues:
1. If restyle a map, the thumbnail doesn't change.
2. Performance for multiple file uploads is really slow on dataset save and dataset page load.

**Comment by landreev** [comment]
Checked in some fixes: 
1. the worldmap thumbnail should be getting updated properly; I've been having trouble testing it, since WorldMap keeps choking... but hopefully it'll come back up eventually. 
2. This one should have been in in the "general swift support" issue... but I've checked in an optimization fix that should make the dataset page load noticeably faster. If there are any further issues with general swift performance, not explicitly related to the derivative files - let's handle it there. 

3. Note that this appeared to be specific to a worldmap preview as well (not just any thumbnail); similarly to 1. I haven't seen this one again, not sure if it was a timing issue... But even if it happens again, the fix for 1. should take care of it - as in, it'll get updated with a legit thumbnail, eventually.
